cmds/core/shasum/shasum.go

```diff
@@ -5,83 +5,80 @@
 package main
 
 import (
+	"bytes"
 	"crypto/sha1"
 	"crypto/sha256"
 	"fmt"
 	"io"
+	"log"
 	"os"
 
 	"github.com/spf13/pflag"
 )
 
-func helpPrinter() {
-	fmt.Printf("Usage:\nshasum -a <algorithm> <File Name>\n")
-	pflag.PrintDefaults()
-	os.Exit(0)
-}
+var (
+	algorithm = pflag.IntP("algorithmhttps://github.com/u-root/u-root/pull/2319", "a", 1, "SHA algorithm, valid args are 1 and 256")
+	help      = pflag.BoolP("help", "h", false, "Show this help and exit")
+)
 
-func versionPrinter() {
-	fmt.Println("shasum utility, URoot Version.")
-	os.Exit(0)
-}
+var usage = "Usage:\nshasum -a <algorithm> <File Name>"
 
-func getInput(fileName string) (input []byte, err error) {
-	if fileName != "" {
-		return os.ReadFile(fileName)
-	}
-	return io.ReadAll(os.Stdin)
+func helpPrinter() {
+	fmt.Println(usage)
+	pflag.PrintDefaults()
 }
 
 //
 // shaPrinter prints sha1/sha256 of given data. The
 // value of algorithm is expected to be 1 for SHA1
 // and 256 for SHA256
 //
-func shaPrinter(algorithm int, data []byte) string {
-	var sha string
-	if algorithm == 256 {
-		sha = fmt.Sprintf("%x", sha256.Sum256(data))
-	} else if algorithm == 1 {
-		sha = fmt.Sprintf("%x", sha1.Sum(data))
+func shaPrinter(w io.Writer, data []byte) error {
+	if *algorithm == 256 {
+		fmt.Fprintf(w, "%x", sha256.Sum256(data))
+	} else if *algorithm == 1 {
+		fmt.Fprintf(w, "%x", sha1.Sum(data))
 	} else {
-		fmt.Fprintf(os.Stderr, "Invalid algorithm")
-		return ""
+		return fmt.Errorf("invalid algorithm")
 	}
-	return sha
+	return nil
 }
 
-func main() {
-	var (
-		algorithm int
-		help      bool
-		version   bool
-	)
-	cliArgs := ""
-	pflag.IntVarP(&algorithm, "algorithm", "a", 1, "SHA algorithm, valid args are 1 and 256")
-	pflag.BoolVarP(&help, "help", "h", false, "Show this help and exit")
-	pflag.BoolVarP(&version, "version", "v", false, "Print Version")
-	pflag.Parse()
-
-	if help {
+func shasum(w io.Writer, r io.Reader, args ...string) error {
+	if *help {
 		helpPrinter()
+		return nil
 	}
 
-	if version {
-		versionPrinter()
-	}
-	if len(pflag.Args()) == 1 {
-		cliArgs = pflag.Args()[0]
-	}
-	input, err := getInput(cliArgs)
-	if err != nil {
-		fmt.Println("Error getting input.")
-		os.Exit(-1)
-	}
-	fmt.Printf("%s ", shaPrinter(algorithm, input))
-	if cliArgs == "" {
-		fmt.Printf(" -\n")
+	buf := &bytes.Buffer{}
```

Comment:
time for me to repeat my usual question: what happens if the file is larger than all of memory? This will fail. 

This is from the docs for shasum

package main

import (
	"crypto/sha256"
	"fmt"
	"io"
	"log"
	"os"
)

func main() {
	f, err := os.Open("file.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		log.Fatal(err)
	}

	fmt.Printf("%x", h.Sum(nil))
}
note it need not read in the whole file at once.
io.Copy copies 8k at a time.

Reply:
Okay io.Copy is removed. Should be fine now. If not, can you do a concrete proposal what you imagine how to solve this problem?